lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using Org.Apache.REEF.Common.Events;
+using Org.Apache.REEF.IO.PartitionedData;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Diagnostics;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+{
+    internal class DataLoadingContext<T> : IObserver<IContextStart>
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(DataLoadingContext<T>));
+
+        [Inject]
+        internal DataLoadingContext(IInputPartition<T> partition)
+        {
+            Console.WriteLine("ENTERED CONTEXT");
+        }
+
+        public void OnNext(IContextStart value)
+        {
```

Comment:
Let me first identify the two points of view. The question is when to trigger the download of the partition information. The two options discussed are:

  * In the constructor of the `IPartition` in question. That way, the data download would be triggered in this new `ContextStart` handler.
  * In the `GetHandle()` method. That way, the data is only downloaded if and when required.

I don't think either of those is particularly attractive, as they hide the intent of data caching (or localization) from the client code. Hence, I propose to add a new method `IInputPartition.Cache()` which performs the download. Ideally, that method would be async and idempotent. That way, it can be called in the context start handler *and* the Task. The context start handler would start the download (in the case of file-backed partitions) and the task would wait for the download to complete before running. If all goes well, the wait time at that point is 0.

Looking at Spark, it is probably also a good idea to add a parameter to `.Cache()` that indicates the level in the memory hierarchy we want the data to go into, e.g. `RAM`, `SSD`, `HDD`.

This cache functionality should probably be its own JIRA. If you agree, I suggest to move forward with this one by leaving the download in `GetHandle()` which is inefficient, but seems to cause less code changes we have to undo in the `.Cache()` scheme.

WDYT?